### ISSUE TITLE
Missing parameters in cadence-innovus-place-route

### ISSUE THREAD

**Issue by hchenji** [mflowgen/mflowgen#91]
cell_padding, useful_skew, useful_skew_ccopt_effort, hold_target_slack (and probably more) from other steps are missing in configure.yml of cadence-innovus-place-route. Building commercial.py flow for gcdunit fails as a result since innovus cannot find these params when it encounters $::env() in the tcl scripts.

**Comment by ctorng** [comment]
Thanks for this issue!

Right... the big place-route node was derived from the full flow (the one with all the nodes split out). I think the full flow has drifted away, and some maintenance has to be done to bring this single-step place-route node back.

Honestly, I think the place-route node should be tremendously simplified...

- No longer be done with Innovus Foundation Flow
- Simple understandable "minimal working example" script
- Most people who want a single place-route node just want rough numbers and not tapeout-quality numbers

Does that match what you are actually interested in?

**Comment by hchenji** [comment]
Yes that's exactly what I was looking for - a simple way to get reasonable accurate power/area/timing numbers with reasonable default values. This is very useful especially when you're a non-expert who just wants to evaluate simple hardware designs and get some rough numbers. Thanks again for this project, saves me a lot of time.

**Comment by ctorng** [comment]
We'll see what we can do.

**Comment by hchenji** [comment]
Will do.

Question: for the RTL step, how can we combine multiple .v files into a single design.v? I see that synopsys-dc-synthesis is hardwired to take in a single .v file, but I'm looking at a fairly small project with 10 different system verilog files, including some interfaces...

**Comment by ctorng** [comment]
Assuming there are no module name clashes, SVerilog files can all be directly cat'ed together.

Right, the common library synopsys-dc-synthesis node expects a single .v file. We preferred this because:

- Many designs are simple without many files
- Many newer hardware generation languages (e.g., Chisel, PyMTL, Magma, HLS tools) can produce a single file with everything in it

In addition, it is often nice when the "rtl" node outputs exactly what the design is without references flying around.

If that is not what you want, then the mflowgen way to get something different is:

- Unplug the synthesis node and plug in your own version (this always works because you can do whatever you actually want in your custom node)
- Or you can take advantage of submodularity (i.e., how we designed the node to be tweaked)
    - https://mflowgen.readthedocs.io/en/latest/stdlib-submodular-nodes.html

Submodularity falls into the category of "clever node design". The goal is that you can override small pieces in a node while reusing the rest of it as is. So in your case, you would override the "read-design.tcl" with your own version (create a custom node with your version and connect it to synthesis as an input). Right now, submodularity is implemented with a tcl script runner (see START.tcl in the node) and is a bit clunky, but it allows to reuse the rest of the node as is.
